### ISSUE TITLE
Support for Fenghuang

### ISSUE THREAD

**Issue by lunzima** [irusanov/ZenStates#2]
I have a AMD Fenghuang (K17.5) and it's not supported by ZenStates.
![CPU-Z](https://user-images.githubusercontent.com/32062239/70904918-8a15cd80-203d-11ea-9de8-e2c13d7e0388.png)

**Comment by lunzima** [comment]
Trying with CPUHandler.CPUType.DEBUG and it throws InvalidArguments exception
![Uploading TIM截图20191216201243.png…]()

**Comment by irusanov** [comment]
Try it with forcing CPUtype to Raven_Ridge. DEBUG type doesn't do much and I haven't changed that part of the code. It only sets some Pstates and default values, but nothing more. You can't use most of the functions, because it needs correct addresses.

**Comment by lunzima** [comment]
It seems that this CPU has only two P-state P0 and P1,
ZenStates can read P0 correctly but not P1.
![TIM截图20191217214016](https://user-images.githubusercontent.com/32062239/71000240-e9dfa780-2115-11ea-8c6d-84856478765d.png)

**Comment by irusanov** [comment]
Thanks for the report. This makes it harder with all the variations. I will have to think a bout a way to detect supported P-States reliably.

**Comment by irusanov** [comment]
First version is up: https://github.com/irusanov/SMUDebugTool
You can set the addresses and command manually. There's basic support for Renoir, however I don't know if it works at all.

**Comment by lunzima** [comment]
Got nothing with SMUDebugTool.
![TIM截图20200107220800](https://user-images.githubusercontent.com/32062239/71901124-33159b00-319a-11ea-80f0-f596c3fa5e60.png)
An outsourced AMD field application engineer (literally "Agricultural Enterprise contract employee") said this is expected.

**Comment by irusanov** [comment]
Yes, that's expected. It probably has different offset between msg and rsp. I will give you a set if addresses to try.

**Comment by irusanov** [comment]
Can you try these sets:
```
CMD Address: 0x3B1051C
RSP Address: 0x3B10568
ARG Address: 0x3B10590
```
```
CMD Address: 0x3B10A20
RSP Address: 0x3B10A80
ARG Address: 0x3B10A88
```
```
CMD Address: 0x3B10524
RSP Address: 0x3B10570
ARG Address: 0x3B10590
```
```
CMD Address: 0x3B10524
RSP Address: 0x3B10570
ARG Address: 0x3B10A40
```
```
CMD Address: 0x3B10520
RSP Address: 0x3B1056C
ARG Address: 0x3B109B8
```

Try to send 0x1 and/or 0x2 command from SMU tab and if any of these give you OK, then you can use the addresses that worked. If not, we'll have to try something else.

**Comment by lunzima** [comment]
![TIM截图20200109203051](https://user-images.githubusercontent.com/32062239/72068100-371ef580-331f-11ea-837f-47950cb60744.png)
The second set of SMU address is working, so you should treat Fenghuang as a variant of Raven although useful mailbox commands is limited.
With wrong SMU address the graphics driver goes wrong and produces garbled display, so be careful.

**Comment by irusanov** [comment]
Thanks for the valuable info. Yes, the tool is not entirely safe and cause unpredictable issues.
Based on your test, it seems to be using Picasso's SMU addresses. That is Zen+ with Vega graphics.

The commands for the APUs are entirely different than the ones of the desktop parts. It is more in line with what FlyGoat has in his repo, but I have to compare. There are commands for the GFX on top of the CPU ones and I think it needs a little different approach.

Perhaps it's a good idea to have a separate app for APUs, because it would be different to support all possible CPUs in one package.

**Comment by lunzima** [comment]
Flygoat (aka "Bed sheet contract employee") has been implementing "Core Boost" functions on Loongson 3A4000 (with integrated GS132 "SMU" equivalent) these days, so there might be few update on his ryzen_nb_smu :-(

**Comment by alexantone** [comment]
> Thanks for the report. This makes it harder with all the variations. I will have to think about a way to detect supported P-States reliably.

Based on the amd MSR documentation the number of pstates is given by bits [6:4] of
MSR 0xC0010061 [P-state Current Limit]*
Bits | Description
----- | -----------
63:7  | Reserved. 
6:4    |PstateMaxVal: P-state maximum value. Specifies the lowest-performance non-boosted P-state (highest non-boosted value) allowed.
3       | Reserved. Read-only. Reset: Fixed,0. 
2:0    | CurPstateLimit: current P-state limit. Specifies the highest-performance P-state (lowest value) allowed. CurPstateLimit is always bounded by PstateMaxVal.

*[Open-Source Register Reference for AMD Family 17h Processors (PUB)](https://developer.amd.com/wp-content/resources/56255_3_03.PDF)

**Comment by irusanov** [comment]
Yes, I've have it implemented in the new version of the app some time ago. There is more information available, discovered just recently and I will try to work on it in the next weeks.

**Comment by irusanov** [comment]
This test version should support it now.

[ZenStates_2.0.0_debug_20200529.zip](https://github.com/irusanov/ZenStates/files/4699573/ZenStates_2.0.0_debug_20200529.zip)

**Comment by lunzima** [comment]
Sorry, but it does not work.
[
![TIM截图20200611215942](https://user-images.githubusercontent.com/32062239/84393971-e180a900-ac2e-11ea-9511-30b05578a505.png)
](url)

**Comment by irusanov** [comment]
I mean the P-States. Should detect just 2 P-States which could be manipulated. The manual OC will be disabled or removed for CPUs that don't support it.
